runtime/lua/vim/lsp/completion.lua

```diff
@@ -0,0 +1,717 @@
+      responses[client_id] = { err = err, result = result }
+      remaining_requests = remaining_requests - 1
+      if remaining_requests == 0 then
+        callback(responses)
+      end
+    end, bufnr)
+
+    if ok then
+      request_ids[client_id] = request_id
+    end
+  end
+
+  return function()
+    for client_id, request_id in pairs(request_ids) do
+      local client = lsp.get_client_by_id(client_id)
+      if client then
+        client.cancel_request(request_id)
+      end
+    end
+  end
+end
+
+--- @param handle vim.lsp.completion.BufHandle
+local function insert_char_pre_cb(handle)
+  if tonumber(vim.fn.pumvisible()) == 1 then
+    if Context.isIncomplete then
+      reset_timer()
+
+      local debounce_ms = next_debounce()
+      if debounce_ms == 0 then
+        vim.schedule(M.trigger)
+      else
+        completion_timer = new_timer()
+        completion_timer:start(debounce_ms, 0, vim.schedule_wrap(M.trigger))
+      end
+    end
+
+    return
+  end
+
+  local char = api.nvim_get_vvar('char')
+  if not completion_timer and handle.triggers[char] then
+    completion_timer = assert(vim.uv.new_timer())
+    completion_timer:start(25, 0, function()
+      reset_timer()
+      vim.schedule(M.trigger)
+    end)
+  end
+end
+
+local function insert_leave_cb()
+  reset_timer()
+  Context.cursor = nil
+  Context:reset()
+end
+
+local function complete_done_cb()
+  local completed_item = api.nvim_get_vvar('completed_item')
+  if not completed_item or not completed_item.user_data or not completed_item.user_data.nvim then
+    Context:reset()
+    return
+  end
+
+  local cursor_row, cursor_col = unpack(api.nvim_win_get_cursor(0)) --- @type integer, integer
+  cursor_row = cursor_row - 1
+  local completion_item = completed_item.user_data.nvim.lsp.completion_item --- @type lsp.CompletionItem
+  local client_id = completed_item.user_data.nvim.lsp.client_id --- @type integer
+  if not completion_item or not client_id then
+    Context:reset()
+    return
+  end
+
+  local bufnr = api.nvim_get_current_buf()
+  local expand_snippet = completion_item.insertTextFormat == protocol.InsertTextFormat.Snippet
+    and (completion_item.textEdit ~= nil or completion_item.insertText ~= nil)
+
+  Context:reset()
+
+  local client = lsp.get_client_by_id(client_id)
+  if not client then
+    return
+  end
+
+  local offset_encoding = client.offset_encoding or 'utf-16'
+  local resolve_provider = (client.server_capabilities.completionProvider or {}).resolveProvider
+
+  local function clear_word()
+    if not expand_snippet then
+      return nil
+    end
+
+    -- Remove the already inserted word.
+    local start_char = cursor_col - #completed_item.word
+    local line = api.nvim_buf_get_lines(bufnr, cursor_row, cursor_row + 1, true)[1]
+    api.nvim_buf_set_text(bufnr, cursor_row, start_char, cursor_row, #line, { '' })
+    return line:sub(cursor_col + 1)
+  end
+
+  --- @param suffix? string
+  local function apply_snippet_and_command(suffix)
```

Comment:
I run into the following issue while testing this PR, I think we need to clear the range specified by the completion_item before applying the text edit. These two videos shows behaviour with the current changes and after applying the fix respectively

https://github.com/neovim/neovim/assets/16278108/7ae8cf1e-e455-47f5-9b72-2a39d7be7657

https://github.com/neovim/neovim/assets/16278108/5fdaa83b-2c97-4c5e-99e5-2c801b5bb0e3

The fix I have right now is, not sure how hacky it's :D

```lua
local function apply_snippet_and_command(suffix)
  if completion_item.textEdit then
    vim.lsp.util.apply_text_edits(
      { { newText = '', range = completion_item.textEdit.range } },
      api.nvim_get_current_buf(),
      'utf-16'
    )
  end
  if expand_snippet then
    apply_snippet(completion_item, suffix)
  end
...
```

I had a similar issue when I used `nvim-lsp-compl`, at the time I had a hacky solution by doing https://github.com/mfussenegger/nvim-lsp-compl/compare/master...JafarAbdi:nvim-lsp-compl:fix_textedit

Reply:
I don't fully understand the issue in the first video. Isn't the completion being inserted correctly?
Also can you reproduce the issue without Copilot/auto-closing parentheses or other plugins that might also affect the completion behavior?

Reply:
Sorry for the delayed response, I was debugging the issue. It's more related to clangd when [editsNearCursor](https://reviews.llvm.org/D75739) is enabled, clangd will provide completions that correct . to ->, and vice-versa. I shared a new video with the minimal configs, when I press `<c-y>` line `a.->i` should become `a->i` and `a->.get()` should become `a.get()`

https://github.com/neovim/neovim/assets/16278108/52ab5953-0d17-47e2-a850-635b8944a6ca

Mimimal config to reproduce the issue `nvim --clean -u minimal.lua test.cpp`

```lua
local general_group = vim.api.nvim_create_augroup("GeneralCommands", {})

vim.api.nvim_create_autocmd("LspAttach", {
  callback = function(args)
    vim.lsp.completion.enable(args.data.client_id, args.buf, { autotrigger = true })
    vim.keymap.set("i", "<c-space>", function()
      vim.lsp.completion.trigger()
    end, { buffer = args.buf })
  end,
  group = general_group,
})

vim.api.nvim_create_autocmd("FileType", {
  pattern = "cpp",
  group = general_group,
  callback = function(args)
    local config = {
      name = "clangd",
      cmd = { "clangd", "--completion-style=detailed" },
      handlers = {},
      capabilities = vim.tbl_deep_extend("force", vim.lsp.protocol.make_client_capabilities(), {
        textDocument = {
          completion = {
            editsNearCursor = true,
          },
        },
        offsetEncoding = { "utf-16" },
      }),
      on_attach = function(client, bufnr) end,
      init_options = vim.empty_dict(),
      settings = vim.empty_dict(),
      root_dir = vim.fs.root(args.file, { ".git" }),
    }
    vim.lsp.start(config)
  end,
})
```

Reply:
> textDocument = {
>   completion = {
>     editsNearCursor = true,
>   },
>  },

Hmmm I don't know, since neovim doesn't set this capability it's technically not responsible for supporting it. I also investigated more about this setting in the clang code and it [looks like an LSP extension](https://github.com/llvm/llvm-project/blob/686600b521aa3ca5fb9f015d25636067e2c5c9fa/clang-tools-extra/clangd/Protocol.h#L471-L474). The goal of this PR (and of neovim's LSP support in general) is to stick to the protocol.

Reply:
Let's keep it as is for the initial merge. We can then follow up on this. 
The proposed change itself doesn't look too bad, so if it doesn't break other scenario I'd be open to include it - if it comes with tests and everything.

Reply:
Fair enough, I'll try to get a better fix with tests in a follow-up PR. [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig)'s users will get this behaviour since it sets `editsNearCursor` to `true`

Thanks @MariaSolOs for the PR!

Reply:
@JafarAbdi sounds good. Once this gets merged feel free to create an issue here so that we can track the follow-up work.